lib/formatter.ex

```diff
@@ -0,0 +1,289 @@
+    )
+  end
+
+
+  ## Returns a copy of `opts` with defaults applied
+  @spec normalize_opts(keyword) :: opts
+  defp normalize_opts(opts) do
+    [
+      indent: opts[:indent] || "  ",
+      line_separator: opts[:line_separator] || "\n",
+      record_separator: opts[:record_separator] || opts[:line_separator] || "\n",
+      after_colon: opts[:after_colon] || " ",
+    ]
+  end
+
+
+  ## Returns an iolist containing `depth` instances of `opts[:indent]`
+  @spec tab(opts, non_neg_integer, iolist) :: iolist
+  defp tab(opts, depth, output \\ []) do
+    if depth < 1 do
+      output
+    else
+      tab(opts, depth-1, [opts[:indent] | output])
+    end
+  end
+
+
+  @typep pp_state :: {
+    non_neg_integer,  ## depth -- current nesting depth
+    boolean,          ## in_str -- is the current byte in a string?
+    boolean,          ## in_bs -- does the current byte follow a backslash in a string?
+    boolean,          ## empty -- is the current object or array empty?
+    boolean,          ## first -- is this the first object or array in the input?
+  }
+
+  @spec pp_iodata(
+    iodata,           ## input -- input data
+    iolist,           ## output_acc -- output iolist (built in reverse order)
+    non_neg_integer,  ## depth -- current nesting depth
+    boolean,          ## in_str -- is the current byte in a string?
+    boolean,          ## in_bs -- does the current byte follow a backslash in a string?
+    boolean,          ## empty -- is the current object or array empty?
+    boolean,          ## first -- is this the first object or array in the input?
+    opts
+  ) :: {iolist, pp_state}
+  defp pp_iodata(input, output_acc, depth, in_str, in_bs, empty, first, opts)
+
+  defp pp_iodata("", output_acc, depth, in_str, in_bs, empty, first, opts) do
+    {:lists.reverse(output_acc), {depth, in_str, in_bs, empty, first, opts}}
+  end
+
+  defp pp_iodata([], output_acc, depth, in_str, in_bs, empty, first, opts) do
+    {:lists.reverse(output_acc), {depth, in_str, in_bs, empty, first, opts}}
+  end
+
+  defp pp_iodata(<<byte::size(8), rest::binary>>, output_acc, depth, in_str, in_bs, empty, first, opts) do
+    pp_byte(byte, rest, output_acc, depth, in_str, in_bs, empty, first, opts)
+  end
+
+  defp pp_iodata(byte, output_acc, depth, in_str, in_bs, empty, first, opts) when is_number(byte) do
+    pp_byte(byte, [], output_acc, depth, in_str, in_bs, empty, first, opts)
+  end
+
+  defp pp_iodata(list, output_acc, depth, in_str, in_bs, empty, first, opts) when is_list(list) do
+    starting_state = {depth, in_str, in_bs, empty, first, opts}
+    {reversed_output, end_state} = Enum.reduce list, {[], starting_state}, fn (item, {output_acc, state}) ->
+      {depth, in_str, in_bs, empty, first, opts} = state
+      {item_output, new_state} = pp_iodata(item, [], depth, in_str, in_bs, empty, first, opts)
+      {[item_output | output_acc], new_state}
+    end
+    {[:lists.reverse(reversed_output) | output_acc], end_state}
+  end
```

Comment:
I wonder if manual recursion wouldn't be actually more readable in here:

```elixir
defp pp_iodata([head | tail], output_acc, depth, in_str, in_bs, empty, first, opts) do
  {data_head, {depth, in_str, in_bs, empty, first, opts}} = 
    pp_iodata(head, output_acc, depth, in_str, in_bs, empty, first, opts)

  # this builds in reversed order, like the current implementation
  pp_iodata(tail, [data_head | output_acc], depth, in_str, in_bs, empty, first, opts)
end
```

We could also optimise charlists a bit:
```elixir
defp pp_iodata([head | tail], output_acc, depth, in_str, in_bs, empty, first, opts) 
     when is_integer(head) do
  pp_byte(head, tail, output_acc, depth, in_str, in_bs, empty, first, opts)
end
```

Reply:
Nice, I like this.

Reply:
I spoke too soon.  I like the directly recursive version better for its looks, but it's leading me into a fight with Dialyzer.  I kept the original `Enum.reduce` implementation.